### ISSUE TITLE
Refactoring: convex hulls

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#4289]
As we are currently looking into storing feature maps in SQL we found some old quirks.
Currently, we support a vector of arbitrary hull points + a vector of bounding boxes.
The only use case for arbitrary hull points in OpenMS is storing mass traces (mainly for debug purposes).
I propose to only store the vector of bounding boxes and think about a mechanism to store arbitrary mass traces.

**Comment by hroest** [comment]
Sounds good, if you need to store mass traces as we do and FFMetabo there
is dedicated data structures such as MSChromatogram and MassTrace

**Comment by hroest** [comment]
>  I guess these just correspond to mass traces of chromatograms so I am currently thinking about whether we should add both an (in most cases empty) vector of chromatograms & a vector of masstraces to each feature.

both sounds like a waste - what about adding a ptr and only when the feature is used it consumes memory? (except the 8 bytes for the ptr)

**Comment by timosachsenberg** [comment]
Yes thats possible

**Comment by hroest** [comment]
> I guess these just correspond to mass traces of chromatograms

yes they do

**Comment by cbielow** [comment]
> only store a set of bounding boxes
why a set? the number of items will be small. Use vector (performance)

I'm not sure about the best design for mass traces. The most efficient way to store this optional information would be inside the `FeatureMap`, as a vector `m_` which has the same length as the feature vector or empty (if features do not have mass traces). However, then there is no way to directly request the MTs given a feature. You'd always need the map.
If the current code would allow this without much hassle, I'd go for that. Otherwise, unique_ptr as Hannes suggested.

**Comment by hendrikweisser** [comment]
> why a set? the number of items will be small. Use vector (performance)

I agree with this. If you don't need ordering or uniqueness, there's no good reason for a set.

> I'm not sure about the best design for mass traces.

I would question whether they need to be stored in a feature map at all. For debugging/visualisation, couldn't they go in an optional mzML output (as chromatograms)?

**Comment by timosachsenberg** [comment]
Yeah probably better to just reference the mass traces.
The problem is that it is already used at quite some places for e.g., fitting a model to the points in the convex hull (= as mass trace).
Don't know if I will find time to refactor that now.

**Comment by hroest** [comment]
>  For debugging/visualisation, couldn't they go in an optional mzML output (as chromatograms)?

thats what we currently do in OpenSWATH
